A predicate to validate all pointer derefs in an expression AST

SaswatPadhi opened the pull request:
This is useful for:
+ checking containment in assigns clause verification
+ havocing only valid memory locations in assigns clause replacement
+ havocing only valid memory locations in loop contracts application
+ creating history snapshots only for valid `old` expressions

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a ~~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

SaswatPadhi committed:
Refactor code contracts utility functions

In this change, we move two of the "static" functions within
contracts.cpp to utils.cpp (a new file in contracts module).

We also fix the #include dependence -- assigns should not #include
contracts.h, it should be the other way round.

kroening commented:
This should not go into `util/` until there is prospect for at least 3 users.

feliperodri commented on regression/contracts/assigns_enforce_structs_07/main.c:
```diff
@@ -28,6 +28,7 @@ int main()
   struct pair *p = nondet_bool() ? malloc(sizeof(*p)) : NULL;
   f1(p);
   struct pair_of_pairs *pp = malloc(sizeof(*pp));
+  pp->p = nondet_bool() ? malloc(sizeof(*(pp->p))) : NULL;
```
Shall we use `--malloc-may-fail` and `--malloc-fail-null` in this test case instead?

SaswatPadhi replied:
Good idea. Done.

feliperodri commented on src/goto-instrument/contracts/utils.cpp:
```diff
@@ -0,0 +1,67 @@
+
+#include "utils.h"
+
+// Create a lexicographic less-than relation between two tuples of variables.
+// This is used in the implementation of multidimensional decreases clauses.
```
This should go in the header file.

feliperodri commented on src/goto-instrument/contracts/utils.h:
```diff
@@ -0,0 +1,31 @@
+
+#include <goto-programs/goto_model.h>
+
+/// \brief
+/// Create an expression that represents a lexicographic less-than
+/// comparison between two tuples of variables.
+/// This is used in instrumentation of decreases clauses.
```
Complete the header.
```c
/// \brief ...
///
/// Create a lexicographic less-than relation between two tuples of variables.
/// This is used in the implementation of multidimensional decreases clauses.
///
/// \param lhs ...
/// \param rhs ...
/// \return ...
```

jimgrundy replied:
+1

SaswatPadhi replied:
Done.

feliperodri commented on src/goto-instrument/contracts/utils.h:
```diff
@@ -0,0 +1,31 @@
+  const std::vector<symbol_exprt> &lhs,
+  const std::vector<symbol_exprt> &rhs);
+
+void insert_before_swap_and_advance(
```
Let's add a header comment here too.

jimgrundy replied:
+1

SaswatPadhi replied:
Done.

feliperodri commented on src/goto-instrument/havoc_utils.cpp:
```diff
@@ -27,47 +29,37 @@ void append_safe_havoc_code_for_expr(
-      goto_programt::make_assignment(expr, std::move(rhs), source_location));
-    t->code_nonconst().add_source_location() = source_location;
-  });
+  side_effect_expr_nondett rhs(expr.type(), source_location);
```
IF this function is specific for scalars, we should add a precondition to make sure `expr` is a scalar.

feliperodri replied:
@SaswatPadhi ?

SaswatPadhi replied:
This function is an internal one, and any failures here should raise exceptions as opposed to user-friendly errors.

Given that, I am not sure if we need such a check here -- the `side_effect_expr_nondett rhs(expr.type(), location);` call on line 61 would fail if the `expr.type()` is not a scalar type.

feliperodri commented on src/util/pointer_predicates.cpp:
```diff
@@ -205,3 +205,17 @@ exprt object_lower_bound(
   return binary_relation_exprt(std::move(p_offset), ID_lt, std::move(zero));
 }
+
+exprt validate_all_derefs(const exprt &expr, const namespacet &ns)
```
`s/validate_all_derefs/validate_all_dereferences/g`

feliperodri replied:
We should move this to `contracts/utils.cpp`.

SaswatPadhi replied:
Done.

feliperodri commented on src/util/pointer_predicates.h:
```diff
@@ -50,6 +50,8 @@ exprt object_upper_bound(
   const exprt &pointer,
   const exprt &access_size);
 
+exprt validate_all_derefs(const exprt &ptr, const namespacet &);
```
Add a comment block.

feliperodri replied:
We should move this to `contracts/utils.h`.

feliperodri replied:
`s/validate_all_derefs/validate_all_dereferences/g`

SaswatPadhi replied:
Done.

feliperodri reviewed (changes requested):
Only minor comments.

jimgrundy commented on src/goto-instrument/contracts/utils.cpp:
```diff
@@ -0,0 +1,67 @@
+  {
+    binary_relation_exprt component_i_equality{lhs[i], ID_equal, rhs[i]};
+    equality_conjunctions[i] =
+      and_exprt(equality_conjunctions[i - 1], component_i_equality);
```
The lack of test coverage reported for this loop body seems concerning.

jimgrundy commented on src/goto-instrument/havoc_utils.cpp:
```diff
@@ -92,3 +84,29 @@ void append_havoc_code(
+        append_safe_havoc_code_for_expr(source_location, ns, expr, dest, [&]() {
+          append_object_havoc_code_for_expr(source_location, expr, dest);
+        });
+        continue;
```
Lack of test coverage on this conditional is a concern

SaswatPadhi replied:
Good point. We do have a test case for multi-dimensional decreases clauses, and it seems to be working as expected. I can take a closer look at it, but I will probably split any fixes (if needed) into a separate PR, and keep this one only about validity checking.

SaswatPadhi committed:
This is useful for:
+ checking containment in assigns clause verification
+ havocing only valid memory locations in assigns clause replacement
+ havocing only valid memory locations in loop contract application
+ creating history snapshots only for valid `old` expressions

feliperodri reviewed (approved):
LGTM.

peterschrammel commented on regression/contracts/assigns_replace_07/main.c:
```diff
@@ -4,22 +4,22 @@
 
 struct pair
```
⛏️  That's not really a `pair` anymore.

SaswatPadhi replied:
Nice catch! Thanks, will fix it + push a couple more minor changes and then merge.

SaswatPadhi replied:
Fixed.
